proposal/0000-record-dot-syntax.md

```diff
@@ -12,77 +12,276 @@ the link, and delete this bold sentence.**
+
+An implementation of this proposal has been battle tested and hardened over 18 months in the enterprise environment as part of [Digital Asset](https://digitalasset.com/)'s [DAML](https://daml.com/) smart contract language (a Haskell derivative utilizing GHC in its implementation), and also in a [Haskell preprocessor and a GHC plugin](https://github.com/ndmitchell/record-dot-preprocessor/). When initially considering Haskell as a basis for DAML, the inadequacy of records was considered the most severe problem, and without devising the scheme presented here, we wouldn't be using Haskell. The feature enjoys universal popularity with users.
 
 ## Proposed Change Specification
 
-Specify the change in precise, comprehensive yet concise language. Avoid words
-like "should" or "could". Strive for a complete definition. Your specification
-may include,
+For this extension we focus on the changes to the parsing rules, and the desugaring, with the expectation the type checking and renamer follow directly from that. To confirm these changes integrate as expected we have written [a prototype implementation](https://gitlab.haskell.org/shayne-fletcher-da/ghc/commits/record-dot-syntax) that parses and desugars the forms directly in the parser.
+
+### `RecordDotSyntax` language extension
+
+This change adds a new language extension (enabled at source via `{-# LANGUAGE RecordDotSyntax #-}` or on the command line via the flag `-XRecordDotSyntax`).
+
+When `RecordDotSyntax` is in effect, the use of '.' to denote function composition is disambiguated from the use of '.' to denote record field access by the presence (, respectively absence,) of whitespace trailing the '.'.
+
+Suppose the following datatype declarations.
+
+```haskell
+data Foo = Foo {foo :: Bar}
+data Bar = Bar {bar :: Baz}
+data Baz = Baz {baz :: Quux}
+data Quux = Quux {quux :: Int}
+```
+
+The existence of the builtin `HasField` typeclass means that it is possible to write code for getting and setting record fields like this:
+
+```haskell
+getQuux :: Foo -> Int
+getQuux a = getField @"quux" (getField @"baz" (getField @"bar" (getField @"foo" a)))
+
+setQuux :: Foo -> Int -> Foo
+setQuux a i = setField@"foo" a (setField@"bar" (getField @"foo" a) (setField@"baz" (getField @"bar" (getField @"foo" a)) (setField@"quux" (getField @"baz" (getField @"bar" (getField @"foo" a))) i)))
+```
+
+`RecordDotSyntax` enables new concrete syntax so that the following program is equivalent.
+
+```haskell
+getQuux a = a.foo.bar.baz.quux
+setQuux a i = a{foo.bar.baz.quux = i}
+```
+
+In the event the language extension is enabled:
+
+| Expression | Equivalent |
```

Comment:
I feel like the primitives are off here. I would expect a primitive for applying an arbitrary function not a primitive for applying operators and an argument and then having to use `&` to apply an arbitrary function.

Reply:
I think `&` emerges naturally. One way of looking at `a.foo.bar.baz` is as reverse composition of projection functions - `&` continues the trend.

Reply:
That is entirely reasonable, but can you think of a good syntax for that? I agree with you in principle, but couldn't think of the syntax.

Reply:
Not really sadly, `lens` calls it `%~` or `%=` in the `MonadState` version but that doesn’t seem great.